### PR TITLE
Don't realise lazy coords in netcdf save.

### PR DESCRIPTION
**Pull request by pp-mo**
### Description
Performance testing revealed that the [recent changes to stream all lazy data in netcdf saves](https://github.com/SciTools/iris/pull/4375) didn't quite fix the problem of realising lazy coordinates, as intended
I.E. nc-save is still loading all lazy aux-coords into memory.

So this fixes that, and adds a test to ensure the behaviour.

---
[Consult Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/contributing_pull_request_checklist.html)

### CODE REVIEW

File: lib/iris/fileformats/netcdf.py [hunk 1]
```diff
@@ -1190,9 +1190,9 @@ def write(
 
         # Ensure that attributes are CF compliant and if possible to make them
         # compliant.
-        self.check_attribute_compliance(cube, cube.lazy_data())
+        self.check_attribute_compliance(cube, cube.dtype)
         for coord in cube.coords():
-            self.check_attribute_compliance(coord, coord.points)
```
Comment (pp-mo) [inline lib/iris/fileformats/netcdf.py:1195]: This was the problem line that realised all the points
(though bounds were ok, interestingly)

I think we **_could_** just replace the `points` here with `core_points()`
However, I also realised that `check_attribute_compliance` call doesn't actually _use_ the data array at all, it only uses the _dtype_ of it.
So, I've replaced that throughout, as an alternative to passing a 'safe' version of the data in the problem location.
Unfortunately that does causes a lot of noisy changes in the tests (though very little in functional code).

File: lib/iris/tests/integration/test_netcdf.py [hunk 2]
```diff
@@ -273,10 +273,16 @@ def test_lazy_preserved_save(self):
...
         self.assertTrue(acube.has_lazy_data())
+        # Also check a coord with lazy points + bounds.
+        self.assertTrue(acube.coord("forecast_period").has_lazy_points())
+        self.assertTrue(acube.coord("forecast_period").has_lazy_bounds())
         with self.temp_filename(".nc") as nc_path:
             with Saver(nc_path, "NETCDF4") as saver:
                 saver.write(acube)
+        # Check that cube data is not realised, also coord points + bounds.
         self.assertTrue(acube.has_lazy_data())
+        self.assertTrue(acube.coord("forecast_period").has_lazy_points())
```
Comment (pp-mo) [inline lib/iris/tests/integration/test_netcdf.py:284]: NOTE: this check fails with the older code.

File: lib/iris/tests/unit/fileformats/netcdf/test_Saver__lazy.py [hunk 3]
```diff
@@ -122,3 +122,7 @@ def test_lazy_streamed_bounds(self):
         self.cube.replace_coord(lazy_coord)
         self.save_common(self.cube)
         self.assertTrue(self.store_watch.called)
+
+
+if __name__ == "__main__":
```
Comment (pp-mo) [inline lib/iris/tests/unit/fileformats/netcdf/test_Saver__lazy.py:127]: Just for consistency, so you can run these tests directly.

**Review by trexfeathers — APPROVED** [review]
Nice catch @pp-mo

**Pull request merged**